backend/pkg/modules/internal/traceanalyzer/guessableid/guessableid.go

```diff
@@ -0,0 +1,240 @@
+	// Do not add it if already present
+	for _, v := range p.history {
+		if v == value {
+			return false // We did not add it
+		}
+	}
+	p.history[p.i] = value
+	p.i = (p.i + 1) % int(p.maxHistory)
+
+	// We reached the end of the array, and circled back to the start.
+	// We have enough data to do something
+	// Let's at least try to guess what is the type of this data
+	if p.i == 0 {
+		p.enoughData = true
+		p.guessTypeHint()
+	}
+
+	return true
+}
+
+func (p *paramHistory) guessTypeHint() {
+	if !p.enoughData { // Should not happen
+		return
+	}
+
+	typeHintHist := make(map[TypeHint]uint)
+	for _, param := range p.history {
+		// Check if it's a UUID
+		if _, err := uuid.FromString(param); err == nil {
+			typeHintHist[HintUUID]++
+		} else {
+			typeHintHist[HintUnknown]++
+		}
+	}
+
+	// Get the most represented typeHint
+	max := uint(0)
+	typeHint := HintUnknown
+	for k, v := range typeHintHist {
+		if v > max {
+			max = v
+			typeHint = k
+		}
+	}
+
+	if float64(max)/float64(p.maxHistory) > HintTypeThreshold {
+		p.typeHint = typeHint
+	}
+}
+
+func (p *paramHistory) isSimilar() (bool, GuessableReason) {
+	if p.typeHint == HintUUID { // It looks like this is a UUID parameter, let's consider it as non guessable
+		return false, GuessableReason{}
+	}
+	d := p.distance()
+	c := p.compressionRatio()
+
+	// Heuristic of the death. If the distance is close to 1, AND the compression
+	// ratio is "pretty" good, it means that the set is pretty similar
+	if d >= DistanceThreshold && c >= CompressionThreshold {
+		return true, GuessableReason{
+			"distance":              d,
+			"distance_threshold":    DistanceThreshold,
+			"compression":           c,
+			"compression_threshold": CompressionThreshold,
+		}
+	}
+
+	return false, GuessableReason{}
+}
+
+func (p *paramHistory) compressionRatio() float32 {
+	var result float32
+
+	var buf bytes.Buffer
+	zw, err := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	if err != nil {
+		return result
+	}
+
+	b := strings.Join(p.history[:], "")
+	_, err = zw.Write([]byte(b))
+	if err != nil {
+		return result
+	}
+	if err := zw.Close(); err != nil {
+		return result
+	}
+
+	compressedLen := buf.Len() - gzipHeaderLen
+
+	// fmt.Printf("=== Original Length: %d, Compressed: %d ===", len(b), compressedLen)
+
+	return float32(len(b)) / float32(compressedLen)
+}
+
+func (p *paramHistory) distance() float32 {
+	sMatrix := make([][]float32, p.maxHistory)
+	for i := range sMatrix {
+		sMatrix[i] = make([]float32, p.maxHistory) // XXX: Optimize creation the triangular matrix should be pre instantiated in the paramHistory struct
+	}
+	for i, pi := range p.history {
+		for j, pj := range p.history {
+			if i == j {
+				// Only compute inferior triangular part
+				break
+			}
+			res, err := edlib.StringsSimilarity(pi, pj, edlib.OSADamerauLevenshtein)
+			if err != nil {
+				sMatrix[i][j] = 0.0
+			} else {
+				sMatrix[i][j] = res
+			}
+		}
+	}
+
+	// Frobenius Norm
+	var sum float32
```

Comment:
We could use some preexisting math functions so we don't reinvent the wheel

Ex: https://pkg.go.dev/gonum.org/v1/gonum/mat#Norm

Although in Go it's sometimes recommended to use a simple function instead of importing a new package or adding an entire module I think we should use gonum. Since we will be adding gonum anyway for the stats modules and it will also make the code simpler.

If you are still not convinced, at least separate this math in a different function

Reply:
The Frobenius norm is the commented part.
This is not relevant to use a separate library to just do the sum if the elements of the matrix.
Maybe we will go back to that in the future.